sklearn/inspection/_partial_dependence.py

```diff
@@ -107,6 +107,54 @@ def _grid_from_X(X, percentiles, grid_resolution):
     return cartesian(values), values
 
 
+def _grid_from_custom_grid(features, custom_grid):
+    """Generate a cartesian product of the 1-d grids in custom_grid.
+
+    Parameters
+    ----------
+    features : scalar, array-like
+        The features for which the partial dependency should be computed.
+
+    custom_grid : dict
+        A dictionary mapping an element of `features` to an array of values where
+        the partial dependence should be calculated for that feature
+
+    Returns
+    -------
+    grid : ndarray, shape (n_points, n_target_features)
+        A value for each feature at each point in the grid. ``n_points`` is
+        the product of the lengths of the arrays in custom_grid.
+
+    values : list of 1d ndarrays
+        The values with which the grid has been created. The size of each
+        array ``values[j]`` is the size of the grid in custom_grid.
+    """
+
+    if isinstance(features, (str, int, float, bool)):
+        features = [features]
+    if set(features) != custom_grid.keys():
+        raise ValueError(
+            "If custom_grid is specified, its keys must equal the values in features!"
+        )
+    values = []
+    for feature in features:
+        value = custom_grid[feature]
+        if not isinstance(value, np.ndarray):
+            value = np.array(value)
+        if value.ndim != 1:
+            raise ValueError(
+                "Grid for feature {} is not a one-dimensional array. Got {} dimensions"
+                .format(feature, value.ndim)
```

Comment:
It looks like `feature` is actually an index here, not a name, so users will not get the feature name in the error.